File: scripts/flightdeck_new.py
"""flightdeck_new.py — create a new deck artifact with correct frontmatter + naming + regen.

The mechanical "authoring entry": instead of an agent re-deriving where a spec/plan/
incident/checklist/chart goes, what frontmatter it needs, and remembering to regen the
INDEX/cockpit, this stamps it deterministically. Mirrors flightdeck_init.py. Pure stdlib.
"""

import argparse
import datetime
import re
import sys
from pathlib import Path

import flightdeck_index  # sibling in scripts/; reused for regen after create

KIND_FOLDER = {
    "spec": "specs",
    "plan": "plans",
    "incident": "incidents",
    "checklist": "checklists",
    "chart": "references",
    "doc": "docs",
}
WORKFLOW = {"spec", "plan"}
KNOWLEDGE = {"incident", "checklist", "chart", "doc"}
DATELESS = {"doc"}   # always <slug>.md, never date-prefixed (standing reference)
SLUG_RE = re.compile(r"^[a-z0-9-]+$")
_DEFAULT_STATUS = {
    "spec": "idea", "plan": "idea",
    "incident": "active", "checklist": "active", "chart": "active", "doc": "active",
}


def _frontmatter(kind, status, date, summary, implements, when_to_read, applies_to):
    lines = ["---", f"status: {status}"]
    if kind in WORKFLOW:
        if summary:
            lines.append(f"summary: {summary}")
        if status != "idea":            # idea-stage workflow omits last_updated (convention)
            lines.append(f"last_updated: {date}")
        if implements:
            lines.append(f"implements: {implements}")
    else:                               # knowledge — routing fields required
        if summary:
            lines.append(f"summary: {summary}")
        lines.append(f"when_to_read: {when_to_read}")
        lines.append(f"applies_to: [{', '.join(applies_to)}]")
        lines.append(f"last_updated: {date}")
        if kind == "incident":
            lines.append("resolved_by:")     # 空=未根治；填 commit/test = 退役依据
    lines.append("---")
    return "\n".join(lines)


_INCIDENT_BODY = """# {title}

## Signature
- symptom: `<报错原文 / 可观测症状>`
- error_type: <异常类型/错误码 或 —>
- where: <函数/文件/子系统>
- trigger: <什么动作/场景引发>

## 症状/复现

## 根因

## 修法

## Cases
- {date_placeholder} 首次
"""


def _body_scaffold(kind, title):
    if kind == "incident":
        return _INCIDENT_BODY.format(title=title, date_placeholder="YYYY-MM-DD")
    return f"# {title}\n"


def new(deck, kind, slug, title, status=None, summary="", implements=None,
        when_to_read=None, applies_to=None, date=None, regen=True):
    """Create one deck artifact; return its Path. Raises ValueError/FileExistsError."""
    if kind not in KIND_FOLDER:
        raise ValueError(f"unknown kind: {kind!r} (one of: {', '.join(sorted(KIND_FOLDER))})")
    if not slug or not SLUG_RE.match(slug):
        raise ValueError(
            f"illegal slug: {slug!r} — must match ^[a-z0-9-]+$ "
            "(lowercase ascii / digits / hyphens, e.g. new-artifact-entrypoint)"
        )
    if not title:
        raise ValueError("title is required")
    if kind in WORKFLOW and not summary:
        raise ValueError(f"{kind} requires --summary (drives the INDEX row; required by flightdeck_index)")
    if kind in KNOWLEDGE and (not when_to_read or not applies_to):
        raise ValueError(f"{kind} requires --when-to-read and --applies-to (required routing fields)")
    if implements and kind not in WORKFLOW:
        raise ValueError(f"--implements is workflow-only; not valid for kind {kind!r}")

    date = date or datetime.date.today().isoformat()
    status = status or _DEFAULT_STATUS[kind]

    folder = Path(deck) / KIND_FOLDER[kind]
    filename = f"{slug}.md" if (status == "idea" or kind in DATELESS) else f"{date}-{slug}.md"
    path = folder / filename
    if path.exists():
        raise FileExistsError(f"artifact already exists: {path}")

    fm = _frontmatter(kind, status, date, summary, implements, when_to_read, applies_to)
    body = _body_scaffold(kind, title)
    path.write_text(f"{fm}\n\n{body}", encoding="utf-8")

    if regen:
        flightdeck_index.main([str(deck)])
    return path


def main(argv=None):
    ap = argparse.ArgumentParser(
        description="Create a new flightdeck deck artifact (frontmatter + naming + INDEX/cockpit regen)."
    )
    ap.add_argument("deck", help="path to the flightdeck/ deck root")
    ap.add_argument("kind", choices=sorted(KIND_FOLDER))
    ap.add_argument("--slug", required=True, help="filename slug: ^[a-z0-9-]+$")
    ap.add_argument("--title", required=True, help="human title (becomes the H1)")
    ap.add_argument("--status", default=None,
                    choices=["idea", "active", "done", "obsolete", "superseded"])
    ap.add_argument("--summary", default="", help="one-line gist (drives the INDEX row)")
    ap.add_argument("--implements", default=None, help="workflow only: specs/<x>.md")
    ap.add_argument("--when-to-read", dest="when_to_read", default=None)
    ap.add_argument("--applies-to", dest="applies_to", default=None, help="comma-separated tags")
    ap.add_argument("--date", default=None, help="YYYY-MM-DD (default: today; override for backfill)")
    args = ap.parse_args(argv)

    applies = [t.strip() for t in args.applies_to.split(",")] if args.applies_to else None
    try:
        path = new(args.deck, args.kind, slug=args.slug, title=args.title,
                   status=args.status, summary=args.summary, implements=args.implements,
                   when_to_read=args.when_to_read, applies_to=applies, date=args.date)
    except (ValueError, FileExistsError) as e:
        print(f"refused: {e}")
        return 2

    status = args.status or _DEFAULT_STATUS[args.kind]
    print(f"created {args.kind} at {path}")
    if args.kind in WORKFLOW and status == "idea":
        print("INDEX updated; cockpit unchanged (status=idea)")
    elif args.kind in WORKFLOW:
        print("INDEX + cockpit (## 进行中) updated")
    else:
        print("folder INDEX updated")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/flightdeck_new.js
#!/usr/bin/env node
'use strict';
// flightdeck_new — create a deck artifact with correct frontmatter + naming +
// INDEX/cockpit regen. Node port of flightdeck_new.py; byte-parity with the
// reference. Built-ins only, zero npm deps.

const fs = require('fs');
const path = require('path');
const idx = require('./flightdeck_index');

const KIND_FOLDER = {
  spec: 'specs', plan: 'plans', incident: 'incidents',
  checklist: 'checklists', reference: 'references', doc: 'docs',
};
const WORKFLOW = new Set(['spec', 'plan']);
const KNOWLEDGE = new Set(['incident', 'checklist', 'reference', 'doc']);
const DATELESS = KNOWLEDGE; // knowledge is always <slug>.md (standing reference, not a log)
const SLUG_RE = /^[a-z0-9-]+$/;
const DEFAULT_STATUS = {
  spec: 'idea', plan: 'idea',
  incident: 'active', checklist: 'active', reference: 'active', doc: 'active',
};

const INCIDENT_BODY = (title) =>
  `# ${title}

## Signature
- symptom: \`<报错原文 / 可观测症状>\`
- error_type: <异常类型/错误码 或 —>
- where: <函数/文件/子系统>
- trigger: <什么动作/场景引发>

## 症状/复现

## 根因

## 修法

## Cases
- YYYY-MM-DD 首次
`;

function frontmatter(kind, status, date, summary, implementsField, whenToRead, appliesTo) {
  const lines = ['---', `status: ${status}`];
  if (WORKFLOW.has(kind)) {
    if (summary) lines.push(`summary: ${summary}`);
    if (status !== 'idea') lines.push(`last_updated: ${date}`);
    if (implementsField) lines.push(`implements: ${implementsField}`);
  } else {
    if (summary) lines.push(`summary: ${summary}`);
    lines.push(`when_to_read: ${whenToRead}`);
    lines.push(`applies_to: [${appliesTo.join(', ')}]`);
    lines.push(`last_updated: ${date}`);
    if (kind === 'incident') lines.push('resolved_by:');
  }
  lines.push('---');
  return lines.join('\n');
}

function bodyScaffold(kind, title) {
  if (kind === 'incident') return INCIDENT_BODY(title);
  return `# ${title}\n`;
}

function vErr(msg) {
  const e = new Error(msg);
  e.code = 'REFUSE';
  return e;
}

function newArtifact(deck, kind, slug, title, opts = {}) {
  const { status = null, summary = '', implements: implementsField = null,
    whenToRead = null, appliesTo = null, date = null, regen = true } = opts;
  if (!(kind in KIND_FOLDER)) {
    throw vErr(`unknown kind: '${kind}' (one of: ${Object.keys(KIND_FOLDER).sort().join(', ')})`);
  }
  if (!slug || !SLUG_RE.test(slug)) {
    throw vErr(`illegal slug: '${slug}' — must match ^[a-z0-9-]+$ ` +
      '(lowercase ascii / digits / hyphens, e.g. new-artifact-entrypoint)');
  }
  if (!title) throw vErr('title is required');
  if (WORKFLOW.has(kind) && !summary) {
    throw vErr(`${kind} requires --summary (drives the INDEX row; required by flightdeck_index)`);
  }
  if (KNOWLEDGE.has(kind) && (!whenToRead || !appliesTo)) {
    throw vErr(`${kind} requires --when-to-read and --applies-to (required routing fields)`);
  }
  if (implementsField && !WORKFLOW.has(kind)) {
    throw vErr(`--implements is workflow-only; not valid for kind '${kind}'`);
  }

  const theDate = date || new Date().toISOString().slice(0, 10);
  const theStatus = status || DEFAULT_STATUS[kind];

  const folder = path.join(deck, KIND_FOLDER[kind]);
  const filename = theStatus === 'idea' || DATELESS.has(kind) ? `${slug}.md` : `${theDate}-${slug}.md`;
  const filePath = path.join(folder, filename);
  if (fs.existsSync(filePath)) throw vErr(`artifact already exists: ${filePath}`);

  const fm = frontmatter(kind, theStatus, theDate, summary, implementsField, whenToRead, appliesTo);
  const body = bodyScaffold(kind, title);
  fs.mkdirSync(folder, { recursive: true });
  fs.writeFileSync(filePath, `${fm}\n\n${body}`, 'utf8');

  if (regen) idx.main([deck]);
  return filePath;
}

function main(argv) {
  const a = { deck: null, kind: null, slug: null, title: null, status: null,
    summary: '', implements: null, whenToRead: null, appliesTo: null, date: null };
  const pos = [];
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i];
    if (t === '--slug') a.slug = argv[++i];
    else if (t === '--title') a.title = argv[++i];
    else if (t === '--status') a.status = argv[++i];
    else if (t === '--summary') a.summary = argv[++i];
    else if (t === '--implements') a.implements = argv[++i];
    else if (t === '--when-to-read') a.whenToRead = argv[++i];
    else if (t === '--applies-to') a.appliesTo = argv[++i];
    else if (t === '--date') a.date = argv[++i];
    else pos.push(t);
  }
  a.deck = pos[0];
  a.kind = pos[1];
  const applies = a.appliesTo ? a.appliesTo.split(',').map((s) => s.trim()) : null;
  let filePath;
  try {
    filePath = newArtifact(a.deck, a.kind, a.slug, a.title, {
      status: a.status, summary: a.summary, implements: a.implements,
      whenToRead: a.whenToRead, appliesTo: applies, date: a.date,
    });
  } catch (e) {
    if (e.code === 'REFUSE') { process.stdout.write(`refused: ${e.message}\n`); return 2; }
    throw e;
  }
  const status = a.status || DEFAULT_STATUS[a.kind];
  process.stdout.write(`created ${a.kind} at ${filePath}\n`);
  if (WORKFLOW.has(a.kind) && status === 'idea') {
    process.stdout.write('INDEX updated; cockpit unchanged (status=idea)\n');
  } else if (WORKFLOW.has(a.kind)) {
    process.stdout.write('INDEX + cockpit (## In Progress) updated\n');
  } else {
    process.stdout.write('folder INDEX updated\n');
  }
  return 0;
}

if (require.main === module) {
  process.exit(main(process.argv.slice(2)));
}

module.exports = { newArtifact, main };
